### PR TITLE
Lets encrypt

### PR DESCRIPTION
**Pull request by mbabauer**
Updates to allow Let's Encrypt Docker Image to be used for auto-generated certificates.

### CODE REVIEW

**Comment by brylie** [comment]
@mbabauer are the conflicts here difficult to resolve?

**Comment by s-devaney** [comment]
+1

**Comment by brylie** [comment]
@madushan1000, may this please be reviewed and/or merged soon?

File: src/modules/meteor/assets/templates/start.sh [hunk 1]
```diff
@@ -29,21 +31,70 @@ docker run \
   <% for(var option in logConfig.opts) { %>--log-opt <%= option %>=<%= logConfig.opts[option] %> <% } %>\
   <% for(var volume in volumes) { %>-v <%= volume %>:<%= volumes[volume] %> <% } %>\
   <% for(var args in docker.args) { %> <%= docker.args[args] %> <% } %>\
+  <% if(typeof sslConfig.autogenerate === "object")  { %> \
+    -e "VIRTUAL_HOST=proclaim.32past8.com" \
```
Comment (brylie) [inline src/modules/meteor/assets/templates/start.sh]: Should `proclaim.32past8.com` be hard coded here?

**Review by brylie — COMMENTED** [review]
What is the `proclaim.32past8.com` string used for here?

**Comment by brylie** [comment]
@mbabauer does this PR auto-renew the LetsEncrypt certificate(s)?

**Comment by came** [comment]
+1

**Comment by mbabauer** [comment]
@brylie The docker image I am using is [here](https://github.com/JrCs/docker-letsencrypt-nginx-proxy-companion).  According to their documentation, it should auto-renew the certs.  It hasn't been 90 days since my instance was started, though, so I can't tell you with 100% certainty.

File: src/modules/meteor/assets/templates/start.sh (hunk 1 above)
Comment (mbabauer) [reply src/modules/meteor/assets/templates/start.sh]: Sorry, that's my host I was using to test with.  I'll update the PR to remove that.

**Comment by kulttuuri** [comment]
+1

**Comment by brylie** [comment]
@mbabauer, conflicts have been resolved and the hardcoded URL was removed. Any timeline for further review of this PR?

**Comment by goooseman** [comment]
there is a little bug.

if there is no ssl section in deploy.js it is throwing an error:

```
[xxx.xxx.xxx.xxx] - Pushing the Startup Script
TypeError ejs:34
    32|   <% for(var volume in volumes) { %>-v <%= volume %>:<%= volumes[volume] %> <% } %>\
    33|   <% for(var args in docker.args) { %> <%= docker.args[args] %> <% } %>\
 >> 34|   <% if(typeof sslConfig.autogenerate === "object")  { %> \
    35|     -e "VIRTUAL_HOST=$HOSTNAME" \
    36|     -e "LETSENCRYPT_HOST=<%= sslConfig.autogenerate.domains %>" \
    37|     -e "LETSENCRYPT_EMAIL=<%= sslConfig.autogenerate.email %>" \

Cannot read property 'autogenerate' of undefined
```

**Comment by brylie** [comment]
@goooseman good catch. @mbabauer would this be hard to fix?

**Comment by goooseman** [comment]
@brylie 
in start.sh change line no 34 to:
```
  <% if(sslConfig && typeof sslConfig.autogenerate === "object")  { %> \
```

testing right now and will make a PR

**Comment by goooseman** [comment]
https://github.com/mbabauer/meteor-up/pull/2

**Comment by goooseman** [comment]
Found one more bug, when library was used without ssl. Fixed. 
https://github.com/mbabauer/meteor-up/pull/3

**Comment by goooseman** [comment]
But with ssl autogenerate object it doesn't work for me.

It just show always 503

**Comment by goooseman** [comment]
f## it. i'm leaving.

what i have done at the end. i installed acmetool manually on the server. he checks for the new cert. and i have made a symlink to the /opt/projectName/config folder:

```
$APP_NAME=test
$HOST=test.discours.io
sudo add-apt-repository ppa:hlandau/rhea
sudo apt-get update
sudo apt-get install acmetool
sudo acmetool quickstart
sudo acmetool want $HOST
mkdir -p /opt/$APP_NAME/config/
rm /opt/$APP_NAME/config/bundle.crt /opt/$APP_NAME/config/private.key
ln -s /var/lib/acme/live/$HOST/fullchain /opt/$APP_NAME/config/bundle.crt
ln -s /var/lib/acme/live/$HOST/privkey /opt/$APP_NAME/config/private.key

```

**Comment by brylie** [comment]
@madushan1000 please review this PR, and advise of any needed changes.

**Comment by RemyLespagnol** [comment]
+1 👍

**Comment by lucbocahut** [comment]
I got this to work by changing line 39 to 
-e "VIRTUAL_HOST=<%= sslConfig.autogenerate.domains %>" \
instead of 
-e "VIRTUAL_HOST=$HOSTNAME" \

**Comment by lucbocahut** [comment]
If its unclear how to get this to work:
`git clone https://github.com/mbabauer/meteor-up.git#lets-encrypt` clones the repo to your current local directory
edit meteor-up/src/modules/meteor/templates/start.sh as indicated above by changing VIRTUAL_HOST=<%= sslConfig.autogenerate.domains %> (will only work if you set autogenerate in mup.js)
`cd meteor-up` cd inside the new directory
`npm install` will create /lib and pull all the required modules
`cd ..` move back out of the directory
`npm install -g meteor-up` install the app from the local directory
you should then be able to use `mup` commands as needed

**Comment by NitroBAY** [comment]
+1

**Comment by NitroBAY** [comment]
Native let's encrypt support has to be the very top priority. Almost every website now need SSL.

**Comment by ilan-schemoul** [comment]
This should be the top priority.

**Comment by ilan-schemoul** [comment]
Great !

**Comment by brylie** [comment]
Correct me if I'm wrong, but we might need to document this feature in the README (#432).

**Comment by ilan-schemoul** [comment]
How do we enable let's encrypt support please ?
EDIT : as npm package hasn't been updated I tried to compile locally the package but I failed (see https://github.com/zodern/meteor-up/issues/98)

**Comment by NitroBAY** [comment]
Does it work like this ?
```
ssh: {
autogenerate: true,
}
```

**Comment by tim-phillips** [comment]
@NitroBAY let's give @zodern more than 16 hours before we complain about the npm package

**Comment by zodern** [comment]
```
meteor: {
  ...
  ssl: {
    autogenerate: {
      email: 'email.address@domain.com',
      domains: 'website.com,www.website.com' // coma separated list of domains
    }
  }
}
```

I will update the npm package and documentation this weekend.

**Comment by ilan-schemoul** [comment]
@babenzele I'm absolutely not complaining I wanted to notice that the compilation fails. Sorry if you think I was in any way aggressive I really think @zodern makes a great work ;)
@zodern thanks a lot, really for your reactivity and all your work

**Comment by zodern** [comment]
The npm package is updated.

**Pull request merged**